application/src/main/java/org/thingsboard/server/service/state/DefaultDeviceStateService.java

```diff
@@ -502,17 +507,21 @@ void updateInactivityStateIfExpired(long ts, DeviceId deviceId, DeviceStateData
         }
     }
 
+    void reportInactivity(long ts, DeviceId deviceId, DeviceStateData stateData) {
+        DeviceState state = stateData.getState();
+        state.setActive(false);
+        state.setLastInactivityAlarmTime(ts);
+        save(deviceId, INACTIVITY_ALARM_TIME, ts);
+        onDeviceActivityStatusChange(deviceId, false, stateData);
+    }
+
     boolean isActive(long ts, DeviceState state) {
         return ts < state.getLastActivityTime() + state.getInactivityTimeout();
     }
 
     @Nonnull
     DeviceStateData getOrFetchDeviceStateData(DeviceId deviceId) {
-        DeviceStateData deviceStateData = deviceStates.get(deviceId);
-        if (deviceStateData != null) {
-            return deviceStateData;
-        }
-        return fetchDeviceStateDataUsingEntityDataQuery(deviceId);
+        return deviceStates.computeIfAbsent(deviceId, this::fetchDeviceStateDataUsingEntityDataQuery);
```

Comment:
remove deviceStates.putIfAbsent(deviceId, deviceStateData); inside fetchDeviceStateDataUsingEntityDataQuery

Reply:
also please rename this method